File: 0641. Design Circular Deque/Solution.py
class MyCircularDeque:

    def __init__(self, k: int):
        self.buff = [0] * k
        self.capacity = k
        self.size = 0
        self.front = k - 1
        self.rear = 0

    def insertFront(self, value: int) -> bool:
        if self.isFull():
            return False
        self.buff[self.front] = value
        self.front = (self.front - 1 + self.capacity) % self.capacity
        self.size += 1
        return True

    def insertLast(self, value: int) -> bool:
        if self.isFull():
            return False
        self.buff[self.rear] = value
        self.rear = (self.rear + 1) % self.capacity
        self.size += 1
        return True

    def deleteFront(self) -> bool:
        if self.isEmpty():
            return False
        self.front = (self.front + 1) % self.capacity
        self.size -= 1
        return True

    def deleteLast(self) -> bool:
        if self.isEmpty():
            return False
        self.rear = (self.rear - 1 + self.capacity) % self.capacity
        self.size -= 1
        return True

    def getFront(self) -> int:
        if self.isEmpty():
            return -1
        return self.buff[(self.front + 1) % self.capacity]

    def getRear(self) -> int:
        if self.isEmpty():
            return -1
        return self.buff[(self.rear - 1 + self.capacity) % self.capacity]

    def isEmpty(self) -> bool:
        return self.size == 0

    def isFull(self) -> bool:
        return self.size == self.capacity


# Your MyCircularDeque object will be instantiated and called as such:
# obj = MyCircularDeque(k)
# param_1 = obj.insertFront(value)
# param_2 = obj.insertLast(value)
# param_3 = obj.deleteFront()
# param_4 = obj.deleteLast()
# param_5 = obj.getFront()
# param_6 = obj.getRear()
# param_7 = obj.isEmpty()
# param_8 = obj.isFull()

File: 0641. Design Circular Deque/Solution.java
class MyCircularDeque {

    private final int[] buff;
    private int front;
    private int rear = 0;
    private int size = 0;

    public MyCircularDeque(int k) {
        buff = new int[k];
        front = k - 1;
    }

    public boolean insertFront(int value) {
        if (isFull()) {
            return false;
        } else {
            buff[front] = value;
            front = (front - 1 + buff.length) % buff.length;
            ++size;
            return true;
        }
    }

    public boolean insertLast(int value) {
        if (isFull()) {
            return false;
        } else {
            buff[rear] = value;
            rear = (rear + 1) % buff.length;
            ++size;
            return true;
        }
    }

    public boolean deleteFront() {
        if (isEmpty()) {
            return false;
        } else {
            front = (front + 1) % buff.length;
            --size;
            return true;
        }
    }

    public boolean deleteLast() {
        if (isEmpty()) {
            return false;
        } else {
            rear = (rear - 1 + buff.length) % buff.length;
            --size;
            return true;
        }
    }

    public int getFront() {
        if (isEmpty()) {
            return -1;
        } else {
            return buff[(front + 1) % buff.length];
        }
    }

    public int getRear() {
        if (isEmpty()) {
            return -1;
        } else {
            return buff[(rear - 1 + buff.length) % buff.length];
        }
    }

    public boolean isEmpty() {
        return size == 0;
    }

    public boolean isFull() {
        return size == buff.length;
    }
}

/**
 * Your MyCircularDeque object will be instantiated and called as such:
 * MyCircularDeque obj = new MyCircularDeque(k);
 * boolean param_1 = obj.insertFront(value);
 * boolean param_2 = obj.insertLast(value);
 * boolean param_3 = obj.deleteFront();
 * boolean param_4 = obj.deleteLast();
 * int param_5 = obj.getFront();
 * int param_6 = obj.getRear();
 * boolean param_7 = obj.isEmpty();
 * boolean param_8 = obj.isFull();
 */
